classes/that.sc

```diff
@@ -0,0 +1,254 @@
+That {
+	classvar <all; // cache/dictionary for all existing instances
+
+	var <name; // is also used to generate Ndef and OSCdef names - must be unique!
+	var <input; // input that gets analyzed
+	var <analyzerFunction; // function which analyzes the input and returns an event
+	var <>callback; // function that will get called with the results as first param
+
+	// public variables
+	var <latestValue; // last stored value/event that was sent from analyzer
+	var <analyzer; // a node proxy which runs the analyser and can be acessed to modify arguments
+
+	// private variables
+	var analyzerResultKeys; // keys of the event that the analyzer returns
+	var defName; // name used for OSCdef and Ndef keys
+	var oscChannelName; // osc channel name to send messages from server to sclang
+	var numInputChannels; // needed for unwrapping multichannel results
+	var oscdef; // responds to messages from analyzer
+
+	*initClass {
+		all = ();
+	}
+
+	*new {|name, input, analyzerFunction, callback|
+		var res = all.at(name);
+		if(res.isNil, {
+			if(analyzerFunction.isNil, {
+				Error("Please provide an analyzerFunction").throw;
+			});
+			callback = callback ? {}; // enforce a function as callback and not allow nil
+			res = super.newCopyArgs(name, input, analyzerFunction, callback).init.();
+			all[name] = res;
+		}, {
+			input !? { res.input = input };
+			analyzerFunction !? { res.analyzerFunction = analyzerFunction };
+			callback !? { res.callback = callback };
+		});
+		^res;
+	}
+
+	init {
+		defName = "that_%".format(name).asSymbol;
+		oscChannelName = "/that/%".format(name);
+		this.input = input;
+	}
+
+	clear {
+		OSCdef(defName).free;
+		Ndef(defName).clear;
+		all[name] = nil;
+	}
+
+	input_ {|newInput|
+		input = newInput;
+		this.prUpdateDefs();
+	}
+
+	analyzerFunction_ {|newAnalyzerFunction|
+		analyzerFunction = newAnalyzerFunction;
+		this.prUpdateDefs();
+
+	}
+
+	prUpdateDefs {
+		this.prCreateNdef();
+		this.prCreateOscDef();
+	}
+
+	prCreateNdef {
+		analyzer = Ndef(defName, {
+			var inputChannels;
+			var analyzerResults;
+			var oscPayload;
+			var summedTriggers;
+
+			inputChannels = input.value.asArray;
+			numInputChannels = inputChannels.size;
+
+			analyzerResults = inputChannels.collect { |inputChannel, i|
+				// is .value(...) some multichannel stuff?
```

Comment:
`value` is a call not on a signal, but on a function that returns a signal graph of UGens. If it specifies the arguments (which is optional), it can make the resulting signal graph depend on them. E.g. it can mix them down to the number of channels passed in, etc.

Reply:
Thanks for the information, wasn't aware that this is a thing in SC. added the exact phrase as a comment for future lurkers or contributors, I hope this is ok for you.